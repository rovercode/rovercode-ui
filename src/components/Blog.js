import React, { useEffect, useState } from 'react';
import { hot } from 'react-hot-loader';
import PropTypes from 'prop-types';
import {
  Grid,
  Typography,
  Card,
  Box,
  Link,
  Paper,
  Button,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import QuestionRender from './QuestionRender';

const useStyles = makeStyles({
  cardcontent: {
    '&:last-child': { paddingBottom: 16 },
  },
  markdown: {
    '& p': { marginBottom: 0, marginTop: 0, overflowWrap: 'anywhere' },
  },
  white: {
    '&.cardroot': {
      backgroundColor: 'white',
    },
    '&.noAnswer': {
      opacity: '70%',
      fontStyle: 'italic',
    },
  },
  gray: {
    '&.cardroot': {
      backgroundColor: '#F9F9F9',
    },
    '&.noAnswer': {
      opacity: '70%',
      fontStyle: 'italic',
    },
  },
  unfinishedRequired: {
    color: 'red',
    '&.cardroot': {
      borderColor: 'red',
    },
  },
});

const Blog = ({
  questions, saveBlogAnswers, isReadOnly, programID,
}) => {
  const [editMode, setEditMode] = useState(false);
  const [blogQuestions, setBlogQuestions] = useState([]);
  const classes = useStyles();

  useEffect(() => {
    setBlogQuestions(questions);
  }, [questions]);

  const toggleEditMode = () => {
    if (editMode) {
      saveBlogAnswers(programID, blogQuestions);
    }
    setEditMode(!editMode);
  };

  const getClass = () => {
    if (!isReadOnly && editMode) {
      return classes.gray;
    }
    return classes.white;
  };

  const getRequiredClass = (required, answer) => {
    if (required && !answer && !isReadOnly) {
      return classes.unfinishedRequired;
    }
    return null;
  };

  return (
    <Grid container direction="column">
      {questions.length === 0 && isReadOnly ? (
        <Typography
          variant="caption"
          style={{ color: 'grey', fontStyle: 'italic' }}
        >
          No Blog Post Available
        </Typography>
      ) : (
        <>
          {!isReadOnly ? (
            <Box mb={2}>
              <Typography
                variant="caption"
                style={{ color: 'grey', fontStyle: 'italic' }}
              >
                *Required
              </Typography>
            </Box>
          ) : null}

          {questions
            .sort((a, b) => a.sequence_number - b.sequence_number)
            .map((item) => (
              <Grid item key={item.id}>
                <Box mb={0.5}>
                  <Typography
                    variant="subtitle2"
                    className={getRequiredClass(item.required, item.answer)}
                  >
                    {item.required ? `${item.question}*` : item.question}
                  </Typography>
                </Box>
                <Box mb={2}>
                  <Card
                    variant="outlined"
                    className={`${getClass()} ${getRequiredClass(
                      item.required,
                      item.answer,
                    )} cardroot`}
                  >
                    <QuestionRender
                      answer={item.answer}
                      required={item.required}
                      questionID={item.id}
                      isReadOnly={isReadOnly}
                      editMode={editMode}
                      toggleEditMode={toggleEditMode}
                      blogQuestions={blogQuestions}
                      getRequiredClass={getRequiredClass}
                      getClass={getClass}
                    />
                  </Card>
                </Box>
              </Grid>
            ))}
          {!isReadOnly ? (
            <>
              <Box
                mb={2}
                style={{
                  display: 'flex',
                  justifyContent: 'space-between',
                  alignItems: 'center',
                }}
              >
                <Link
                  variant="caption"
                  style={{ color: 'grey', textDecoration: 'underline' }}
                  rel="noopener noreferrer"
                  href="https://commonmark.org/help/"
                  target="_blank"
                >
                  Formatting Tips
                </Link>
                <Button color="primary" onClick={toggleEditMode}>
                  {editMode ? 'Save' : 'Edit'}
                </Button>
              </Box>
              <Paper
                style={{
                  backgroundColor: 'rgba(254,173,17,.25)',
                  borderColor: 'rgba(254,173,17,1)',
                  padding: '16px',
                  marginBottom: '16px',
                }}
                variant="outlined"
                padding={2}
              >
                <Typography variant="body1">
                  Blog Posts may be public, do not share personal information!
                  {' '}
                  <Link
                    style={{ color: 'black', textDecoration: 'underline' }}
                    rel="noopener noreferrer"
                    href="https://docs.rovercode.com/privacy"
                    target="_blank"
                  >
                    Learn more.
                  </Link>
                </Typography>
              </Paper>
            </>
          ) : null}
        </>
      )}
    </Grid>
  );
};

Blog.defaultProps = {
  questions: [],
  isReadOnly: false,
};

Blog.propTypes = {
  questions: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      question: PropTypes.string,
      answer: PropTypes.string,
      sequence_number: PropTypes.number,
      required: PropTypes.bool,
    }),
  ),
  saveBlogAnswers: PropTypes.func.isRequired,
  isReadOnly: PropTypes.bool,
  programID: PropTypes.number.isRequired,
};

export default hot(module)(Blog);
