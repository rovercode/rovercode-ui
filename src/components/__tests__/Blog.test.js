import React from 'react';
import { shallow } from 'enzyme';
import { Typography } from '@material-ui/core';
import ReactMarkdown from 'react-markdown';

import Blog from '../Blog';

const questions = [{
  id: 1,
  question: 'Question1',
  answer: null,
  sequence_number: 2,
  required: true,
}, {
  id: 2,
  question: 'Question2',
  answer: null,
  sequence_number: 1,
  required: true,
}, {
  id: 3,
  question: 'Question3',
  answer: 'Answer1',
  sequence_number: 3,
  required: false,
}];

describe('The Blog component', () => {
  test('renders on the page with no errors', () => {
    const wrapper = shallow(<Blog />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Typography).at(0).text()).toBe('No Blog Post Available');
  });

  test('displays questions and answers', () => {
    const wrapper = shallow(<Blog questions={questions} />).dive().dive();

    expect(wrapper.find(Typography).length).toBe(7);
    expect(wrapper.find(Typography).at(0).text()).toBe('*Required');
    expect(wrapper.find(Typography).at(1).text()).toBe('Question2*');
    expect(wrapper.find(Typography).at(2).text()).toBe('No Response');
    expect(wrapper.find(Typography).at(3).text()).toBe('Question1*');
    expect(wrapper.find(Typography).at(4).text()).toBe('No Response');
    expect(wrapper.find(Typography).at(5).text()).toBe('Question3');
    expect(wrapper.find(ReactMarkdown).length).toBe(1);
    expect(wrapper.find(ReactMarkdown).at(0).children().text()).toBe('Answer1');
  });

  test('displays questions and answers in read-only mode', () => {
    const wrapper = shallow(<Blog questions={questions} isReadOnly />).dive().dive();

    expect(wrapper.find(Typography).length).toBe(5);
    expect(wrapper.find(Typography).at(0).text()).toBe('Question2*');
    expect(wrapper.find(Typography).at(1).text()).toBe('No Response');
    expect(wrapper.find(Typography).at(2).text()).toBe('Question1*');
    expect(wrapper.find(Typography).at(3).text()).toBe('No Response');
    expect(wrapper.find(Typography).at(4).text()).toBe('Question3');
    expect(wrapper.find(ReactMarkdown).length).toBe(1);
    expect(wrapper.find(ReactMarkdown).at(0).children().text()).toBe('Answer1');
  });
});
