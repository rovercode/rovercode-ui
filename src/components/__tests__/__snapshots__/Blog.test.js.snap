// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Blog component renders on the page with no errors 1`] = `
<WithStyles(ForwardRef(Grid))
  container={true}
  direction="column"
>
  <WithStyles(ForwardRef(Typography))
    style={
      Object {
        "color": "grey",
        "fontStyle": "italic",
      }
    }
    variant="caption"
  >
    No Blog Post Available
  </WithStyles(ForwardRef(Typography))>
</WithStyles(ForwardRef(Grid))>
`;
